File: app/src/main/java/com/boardgamegeek/ui/dialog/PlayerNumberFilterDialog.java
package com.boardgamegeek.ui.dialog;

import android.content.Context;

import com.boardgamegeek.R;
import com.boardgamegeek.filterer.CollectionFilterer;
import com.boardgamegeek.filterer.PlayerNumberFilterer;

public class PlayerNumberFilterDialog extends SliderFilterDialog {
	@Override
	protected int getCheckboxTextId() {
		return R.string.exact;
	}

	@Override
	protected int getDescriptionId() {
		return R.string.filter_description_player_number;
	}

	@Override
	protected int getAbsoluteMax() {
		return PlayerNumberFilterer.MAX_RANGE;
	}

	@Override
	protected int getAbsoluteMin() {
		return PlayerNumberFilterer.MIN_RANGE;
	}

	@Override
	public int getType(Context context) {
		return new PlayerNumberFilterer(context).getType();
	}

	@Override
	protected CollectionFilterer getPositiveData(Context context, int min, int max, boolean checkbox) {
		return new PlayerNumberFilterer(context, min, max, checkbox);
	}

	@Override
	protected int getTitleId() {
		return R.string.menu_number_of_players;
	}

	@Override
	protected InitialValues initValues(CollectionFilterer filter) {
		int min = PlayerNumberFilterer.MIN_RANGE;
		int max = PlayerNumberFilterer.MAX_RANGE;
		boolean isExact = false;
		if (filter != null) {
			PlayerNumberFilterer data = (PlayerNumberFilterer) filter;
			min = data.getMin();
			max = data.getMax();
			isExact = data.isExact();
		}
		return new InitialValues(min, max, isExact);
	}
}

File: app/src/main/java/com/boardgamegeek/ui/dialog/PlayerNumberFilterDialog.kt
package com.boardgamegeek.ui.dialog

import android.content.Context
import com.boardgamegeek.R
import com.boardgamegeek.filterer.CollectionFilterer
import com.boardgamegeek.filterer.PlayerNumberFilterer
import kotlin.math.roundToInt

class PlayerNumberFilterDialog : SliderFilterDialog() {
    override fun getType(context: Context) = PlayerNumberFilterer(context).type
    override val titleResId = R.string.menu_number_of_players
    override val descriptionResId = R.string.filter_description_player_number
    override val checkboxTextResId = R.string.exact
    override val valueFrom = PlayerNumberFilterer.LOWER_BOUND.toFloat()
    override val valueTo = PlayerNumberFilterer.UPPER_BOUND.toFloat()
    override val supportsNone = false

    override fun initValues(filter: CollectionFilterer?): InitialValues {
        val f = filter as? PlayerNumberFilterer
        return InitialValues(
            (f?.min ?: 4).toFloat(),
            (f?.max ?: 4).toFloat(),
            f?.isExact ?: false,
        )
    }

    override fun createFilterer(context: Context): CollectionFilterer {
        return PlayerNumberFilterer(context).apply {
            min = low.roundToInt()
            max = high.roundToInt()
            isExact = checkboxIsChecked
        }
    }

    override fun describeRange(context: Context): String {
        return (createFilterer(context) as? PlayerNumberFilterer)?.describeRange(" - ").orEmpty()
    }
}
